static/docs/commands-reference/commit.md

```diff
@@ -2,40 +2,212 @@
 
 Record changes to the repository by saving outputs to cache.
 
+## Synopsis
+
 ```usage
     usage: dvc commit [-h] [-q | -v] [-f] [-d] [-R] [targets [targets ...]]
 
     positional arguments:
       targets               DVC files.
-
-    optional arguments:
-      -f, --force           Commit even if checksums for dependencies/outputs
-                            changed.
-      -d, --with-deps       Commit all dependencies of the specified target.
-      -R, --recursive       Commit cache for subdirectories of the specified
-                            directory.
 ```
 
+## Description
+
+Normally DVC commands like `dvc add`, `dvc repro` or `dvc run`, commit the data
+to the DVC cache as the last step. What _commit_ means is that DVC
+
+* Computes a checksum for the file
+* Enters the checksum and file name into the DVC stage file
+* Tells the SCM to ignore the file (e.g. add entry to `.gitignore`).  If the
+  workspace was initialized with no SCM support (`dvc init --no-scm`) this does
+  not happen.
+* Adds the file to the DVC cache
+
+There are many cases where the last step is not desirable. For the DVC commands
+where it is appropriate the `--no-commit` option prevents the last step from
+occurring. The checksum is still computed and added to the DVC file, but the
+file is not added to the cache.
+
+That's where the `dvc commit` command comes into play. It handles that last
+step of adding the file to the DVC cache.
+
+The `dvc commit` command is useful for several scenarios where a stage is in
+development, or takes a long time to run, or must be run on another machine
+with higher compute power. The key is to run DVC commands in a mode where data
+is not immediately committed to the cache (the `--no-commit` or `--no-exec`
+options), and to commit the data as a separate step performed when it is certain
+the data is finalized.
+
+* Code for a stage is under active development, with lots of temporary files
+  sitting in the workspace. Committing those temporary files would clutter up
+  the cache, so it's better to commit them when development is finished.
+* Execution of a stage takes a long time to run, like 24 hours, or requires
```

Comment:
I'm not sure I understand this case. Keep also in mind that `--no-commit` will become the default. So, it's not that it's useful in some strange cases. It's important part of the more explicit workflow - you commit artifacts when you are sure you need them. Usually it's useful in case when you produce a lot of them doing a lot of iterations.

Reply:
This is coming from discussion I saw on Diaspora

That a primary purpose for `--no-commit` as I read the discussion is what I said, to support rapid iteration on code or data without filling the cache with intermediate results.

That's why the one example has the reader do `dvc unprotect` on everything.  According to existing documentation, to edit a file you must `dvc unprotect` before editing.

Reply:
Yes, my question was related to the `Execution of a stage takes a long time to run, like 24 hours, or requires` part. 

As for the `dvc unprotect`. You need it (actually, you won't even need it in the default mode very soon - we'll be using only reflinks or copy cache semantics - to simplify all this stuff) only when you actually want to open the file and edit something inside. You don't need it if you use `dvc run` or `dvc repro` to run/re-run a command. Keep in mind - usually scripts are written in a way they produce files from scratch. So, I would not worry or focus attention on `dvc unprotect` here. I hope this explanation makes sense :)

Reply:
I saw a comment by @ruslan from a couple months ago answering someone asking about running tasks on other hardware and the difficulty of coordinating executing a DVC pipeline stage with getting the results from another machine.  He answered that `--no-commit` was meant for that purpose, and there are other comments talking about long-running computations.   Hence, that's where those statements in the docs came from.

I understand that a typical pipeline will start with a raw data file, normalize and clean up the data, then process it into a result.

Reply:
@efiop do you remember what was that case about? Anyways, I don't feel it's a very general case for the `dvc commit` to be used.

Reply:
@shcheklein Not really, to be honest :slightly_frowning_face: But yeah, that is a suitable workaround for that scenario, not sure if that scenario is general enough though, but I do remember people asking about such stuff a few times and this trick would be useful to note. A more general case would be something like `saving results of your command withing running it again with dvc run`.

Reply:
I still don't quite understand this point. Could you guys explain me how does `dvc commit` help in this case? What is so specific about it? 

`saving results of your command withing running it again with dvc run` - what would be the case? Could you recall a ticker or discrord discussion? And how is it different from the first point in the list?

Reply:
@shcheklein Don't recall tickets to be honest. It pretty much usually looks like so:

User has something long-running and either doesn't want to leave dvc running because it would block all the other dvc commands or is just running his script by himself until he is happy with it. After that he doesn't want to re-run it with `dvc run` again, but he wants to save the results and changes to his pipeline, so he is creating a pipeline stage after-the-fact with `dvc run ... --no-exec` and then using `dvc commit` to just save it as it is without running.

Reply:
It makes sense. @robogeek could we clarify it? This case is not "remote execution" or HPC specific. It similar to `dvc run/repro --no-commit` but when you don't want to use dvc at all (for simplicity and faster iterations as far as I understand, or even if the stage was not created before and it takes a long time to run indeed). We need to clarify all of this better here and in the paragraph before this list.

Reply:
Yes I remember that discussion - the person asked about how to do other DVC commands while another command was running, but was stymied because DVC locks the workspace while it is running a command.  

I can work on it some more

Reply:
I think I have all this covered now.  Look at the latest commit